Add HTTP2 as option in configmap and annotations

TheTechNick opened the pull request:
Hi Nginx Team,

I have added default http2 support to the ingress template.
Is there a reason, that the nginx ingress controller does not enable http2 for ssl connections by default?
The only obstacle I had, was that the environment in the debian container (openssl 1.0.1f) does not support http2. That is why I changed the base container to alpine. (wich should also solve #11)

The changed image is uploaded to: (Tag: 0.5.0)
https://quay.io/repository/nico_schieder/nginxinc-kubernetes-ingress

Please keep in mind that I could not test the nginx plus config.

pleshakov commented:
I don't recommend enabling http/2 by default. It is a new protocol and some browsers still have problems fully implementing it, which leads to problems in some cases.

What we can do is to introduce an option to enable http/2. We can put such an option into [ConfigMap and annotation](https://github.com/nginxinc/kubernetes-ingress/tree/master/examples/customization).

Regarding the base image, we can switch from debian to ubuntu:16.04, which has openssl with support for ALPN required for [Chrome users](https://www.nginx.com/blog/supporting-http2-google-chrome-users/). 
I don't know how many people are using alpine linux so that we can switch to it by default.

We can add support for http/2. Let me know if you're interested in helping.

TheTechNick commented:
Hi,
sure I am interested in helping!
I will make http2 configurable in the configmap and individually on each ingress object via annotations.

Does this look good to you, in terms of annotation naming?
```yaml
annotations:
  nginx.org/enable-http2: 'True'

# configmap
enable-http2: 'True'
```

pleshakov commented:
@TheTechNick thx for the pull request!

The naming of ` nginx.org/http2: 'True'` looks good to me.

> About alpine linux vs. debian/ubuntu:

The arguments you provided for alpine make a good case for alpine. However, I would stick to debian as a default and as an option provide the alpine based image, similar to what we're providing on DockerHub https://hub.docker.com/_/nginx/

> We also have some more pull requests incoming for:
ipv6 binding
log_format
forward secrecy
auth_request
proxy_set_headers Do you prefer small pull requests per option/feature or do you prefer the pull requests to be larger?

That's awesome! Pull requests per option are preferable. If you have an idea how to implement a particular feature, I suggest we discuss it via a GitHub issue

Regarding this pull request:

- Since you're parsing the value of the annotation -- https://github.com/TheTechNick/kubernetes-ingress/blob/914fa992db02e5f353331a14bc287f225a5d1da7/nginx-controller/controller/controller.go#L326 -- it makes sense to report an error in case of a parsing error, similar to https://github.com/nginxinc/kubernetes-ingress/blob/master/nginx-controller/nginx/configurator.go#L198
- Another place where a server is created is here -- https://github.com/TheTechNick/kubernetes-ingress/blob/914fa992db02e5f353331a14bc287f225a5d1da7/nginx-controller/nginx/configurator.go#L147 Http2 must be added
- Let's stick to the debian base image, we will add an option to create an alpine base image. 
- Please remove http2 from the nginx-plus ingress.tmpl.

TheTechNick commented:
> The arguments you provided for alpine make a good case for alpine. However, I would stick to debian as a default and as an option provide the alpine based image, similar to what we're providing on DockerHub https://hub.docker.com/_/nginx/

Sounds great!

> That's awesome! Pull requests per option are preferable. If you have an idea how to implement a particular feature, I suggest we discuss it via a GitHub issue

Watch out for some issues :)
Do I somehow interrupt your plans if I add unittests for some of the config mapping functions?
I would like to be sure new config options do work without having to test them on a  kubernetes cluster.

pleshakov commented:
Thx! Looks good to me!
Before I merge, could you squash all commits into a single commit?

> Do I somehow interrupt your plans if I add unittests for some of the config mapping functions?
I would like to be sure new config options do work without having to test them on a kubernetes cluster.

That would be great if you add unittests for that!

TheTechNick commented:
I have squashed the commits into one and I have also added an entry to examples/customization/nginx-config.yaml, so the configmap entry is documented. Should I also create a examples/http2 folder with the annotations like ssl-services?

pleshakov commented:
Thx! No need for the examples/http2 folder. 

We'll make a new release (0.5.1) this Friday that will include this feature. We'll also create a new tag on our DockerHub to support the alpine distro.

pleshakov commented:
We'll make a new release next week.
